cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -0,0 +1,289 @@
+                                                    long endTime, List<Double> percentiles) throws Exception {
+    List<WorkflowRunRecord> workflowRunRecords = scan(id, startTime, endTime);
+    int count = workflowRunRecords.size();
+
+    if (count == 0) {
+      return null;
+    }
+
+    double avgRunTime = 0.0;
+    for (WorkflowDataset.WorkflowRunRecord workflowRunRecord : workflowRunRecords) {
+      avgRunTime += workflowRunRecord.getTimeTaken();
+    }
+    avgRunTime /= count;
+
+    workflowRunRecords = sort(workflowRunRecords);
+
+    ImmutablePair<Map<String, Long>, Map<String, List<String>>> percentilesAndPercentileList =
+      getPercentiles(workflowRunRecords, percentiles);
+
+    Map<String, Long> percentileToTime = percentilesAndPercentileList.getFirst();
+    Map<String, List<String>> percentileToRunids = percentilesAndPercentileList.getSecond();
+
+    Map<String, List<Long>> actionToRunRecord = getActionRuns(workflowRunRecords);
+
+    Map<String, Map<String, Double>> actionToStatistic = new HashMap<>();
+    for (Map.Entry<String, List<Long>> entry : actionToRunRecord.entrySet()) {
+      double avgForAction = 0;
+      for (long value : entry.getValue()) {
+        avgForAction += value;
+      }
+      avgForAction /= entry.getValue().size();
+      Map<String, Double> actionMap = new HashMap<>();
+      actionMap.put("count", (double) entry.getValue().size());
+      actionMap.put("avgRunTime", avgForAction);
+      actionToStatistic.put(entry.getKey(), actionMap);
+    }
+
+    for (Map.Entry<String, List<Long>> entry : actionToRunRecord.entrySet()) {
+      List<Long> runList = entry.getValue();
+      Collections.sort(runList);
+      for (double percentile : percentiles) {
+        long percentileValue = runList.get((int) ((percentile * runList.size()) / 100));
+        actionToStatistic.get(entry.getKey()).put(Double.toString(percentile), (double) percentileValue);
+      }
+    }
+
+    return new WorkflowStatistics(startTime, endTime, count, avgRunTime, percentileToTime,
+                               percentileToRunids, actionToStatistic);
+
+  }
+
+  private ImmutablePair<Map<String, Long>, Map<String, List<String>>> getPercentiles(
+    List<WorkflowRunRecord> workflowRunRecords, List<Double> percentiles) {
+    int count = workflowRunRecords.size();
+    Map<String, Long> percentileToTime = new HashMap<>();
+    Map<String, List<String>> percentileToRunids = new HashMap<>();
+    for (double i : percentiles) {
+      List<String> percentileRun = new ArrayList();
+      int percentileStart = (int) ((i * count) / 100);
+      for (int j = percentileStart; j < count; j++) {
+        percentileRun.add(workflowRunRecords.get(j).getWorkflowRunId());
+      }
+      percentileToRunids.put(Double.toString(i), percentileRun);
+      percentileToTime.put(Double.toString(i), workflowRunRecords.get(percentileStart).getTimeTaken());
+    }
+    return new ImmutablePair<>(percentileToTime, percentileToRunids);
```

Comment:
it will be better if this is represented in an object, 

``
class PercentileInformation {
double percentile;
long timeTaken;
List<String> runIds;
}
``
and then you can just return a ``List<PercentileInformation>``